### ISSUE TITLE
Can we get 'Targeted Chat Messages?'

### ISSUE THREAD

**Issue by Awesome-Logic** [efroemling/ballistica#32]
Currently only screen_message can be target towards specific clients, it works great but they don't get logged anywhere.

It would specially be useful for servers as we would be able to privately message others which would then get logged in their chat box and reduce spam for other players.

Also, another alternative is to log the message in everyone's chatbox but the popup is limited to specific players.

This would also allow the implementation of personal messages for example by using `@name <message>`

**Comment by efroemling** [comment]
Sure thing; I was mucking with the chat/screen-messaging code earlier so I went ahead and added this. It's in commit c95357e3d0dcf24fb761edd70070a85c69b798c4
_ba.chatmessage() now has an optional 'clients' arg the same as ba.screenmessage(). The only limitation is it only works from the server.

I also replaced the mgr.broadcast() method in the server-wrapper with mgr.chatmessage() and mgr.screenmessage() so you can play with both from there interactively if you want. (it just forwards those calls along to the running binary)

I'm wondering if we'll need a bit more here such as the ability for the server to override the name on the chat message that goes out (so if the server handles forwarding private messages they don't appear to be coming from the server, etc)

**Comment by Awesome-Logic** [comment]
Oh nice, did not know that this was already added.
The name overwrite looks good. A 'PM' label in front of the message could be useful to indicate private messages.

Also if possible, please add a clearscreen command which would be used to clear out all the messages/chats from the main screen.
I currently use a hacky way of doing this: `bs.screenMessage('\n'*10000)`

All of these if used properly would completely eliminate all disturbances during gameplay.

**Comment by efroemling** [comment]
Well, it wasn't added already, but I added it after you suggested it since I was changing that code already :)
I'll try to add the name overwrite and clearscreen commands today...

**Comment by efroemling** [comment]
Ok, I just added a 'sender_override' arg in chatmessage() in the latest commit: d260acebd78ca0902d74d4f3995bcf066b1e4d19

I looked into a screenmessage reset function, but it would require me to change the netplay protocol, which isn't something I want to do until 1.5 is well distributed. So I hope you don't mind doing without that for now.

Let me know if the chatmessage stuff seems to be workable for now and I'll go ahead and close this thread.

**Comment by SushrutShidhaye** [comment]
Can we get option to disable On Screen Left, join and leave Mesaages and also on screen chat messages.
 some players spam server by left joining again and again and send big messages and also after kicking they can use random ids and just irritate us ....
Hope u will help....

**Comment by efroemling** [comment]
Ok I'll look into that stuff and see what is doable easily. (other stuff might have to wait until after 1.5 launch)
I'm wondering if it might make sense to ban IP addresses and not just account IDs... that might at least be harder for people to work around.

**Comment by SushrutShidhaye** [comment]
But using vpn they can join still its a good if its a noob spammer who even dont know about vpn and all...😅

**Comment by Awesome-Logic** [comment]
As much as I know about spamming BS Servers, AccountName (Android1234567 or AwesomeLogic) can be easily spoofed, but on the other hand I've never seen someone manage to spoof their account_id (pb-....).

Currently account_id is only available once the player sends the request to join the lobby.

So there's no way (as much as I know) to get the account_id of a client as soon as they join the server.

Most spam issues could be fixed if _ba.get_game_roster() also gave the account_id of clients as soon as they joined the server, the issue being that there's no way to tell if a client is using spoofed AccountName until it tries to join the lobby.

In Short: Adding a account_id key for every client in _ba.get_game_roster() would fix most spam issues as it is a robust and non-spoofable way of uniquely identifying players. At-least till the day someone manages to spoof account_id like Account Names :/ 
Access to IP Addresses would be cool too. Though that list can currently be made using packet capturing and matching them with join requests. (Another hacky way)

The issue was about 'Targeted Chat Messages' which seem to have been sorted out. This issue can be closed now I guess?

**Comment by SushrutShidhaye** [comment]
Yes, 
But this thing will work only on android id (Android123456) and Google Id(only one google account) because making new pc id and using other google account will change pb- id too(i think).

and as per you suggest he(eric) also need to give more access to servers owners like kick or ban someone from running server using chat commands ( but it will need to kick or ban pb-  id so he cannot join if he change just google id name) and he can join still if he use other google
account...
Or just we can increse left join timing or as i suggest disable on screen left join and leave pop-ups( option to turn on popups or off)...

**Comment by efroemling** [comment]
Hey folks; gonna close this and move the discussion to https://github.com/efroemling/ballistica/issues/41
